generator/mavgen.py

```diff
@@ -64,38 +64,82 @@ def mavgen(opts, args):
             opts.validate = False
 
+
     def expand_includes():
-        """Expand includes in current list of all files, ignoring those already parsed."""
-        for x in xml[:]:
-            for i in x.include:
-                fname = os.path.join(os.path.dirname(x.filename), i)
-
-                # Only parse new include files
-                if fname in all_files:
-                    continue
-                all_files.add(fname)
-
-                # Validate XML file with XSD file if possible.
-                if opts.validate:
-                    print("Validating %s" % fname)
-                    if not mavgen_validate(fname):
-                        return False
-                else:
-                    print("Validation skipped for %s." % fname)
-
-                # Parsing
-                print("Parsing %s" % fname)
-                xml.append(mavparse.MAVXML(fname, opts.wire_protocol))
-
-                # include message lengths and CRCs too
-                x.message_crcs.update(xml[-1].message_crcs)
-                x.message_lengths.update(xml[-1].message_lengths)
-                x.message_min_lengths.update(xml[-1].message_min_lengths)
-                x.message_flags.update(xml[-1].message_flags)
-                x.message_target_system_ofs.update(xml[-1].message_target_system_ofs)
-                x.message_target_component_ofs.update(xml[-1].message_target_component_ofs)
-                x.message_names.update(xml[-1].message_names)
-                x.largest_payload = max(x.largest_payload, xml[-1].largest_payload)
+        """Expand includes. XML files from arguments already parsed inot xml list."""
+        
+        def expand_oneiteration():
+            noincludeadded = True
+
+            for x in xml[:]:
+                try:
+                    if not x.parent:
+                        pass
+                except:
+                    x.parent=set()
+
+                for i in x.include:
+                    fname = os.path.join(os.path.dirname(x.filename), i)
+                    # Only parse new include files
+                    if fname in all_files:
+                        continue
+                    all_files.add(fname)
+                    # Validate XML file with XSD file if possible.
+                    if opts.validate:
+                        print("Validating %s" % fname)
+                        if not mavgen_validate(fname):
+                            print("ERROR Validation of %s failed" % fname)
+                            exit(1)
+                    else:
+                        print("Validation skipped for %s." % fname)
+
+                    # Parsing
+                    print("Parsing %s" % fname)
+
+                    parsed_xml = mavparse.MAVXML(fname, opts.wire_protocol)
+                    #add all parent(s) into included file
+                    parsed_xml.parent=set()
+                    parsed_xml.parent.add(x.basename)
```

Comment:
You appear to be adding apples and oranges to the same set here.  Probably just the basename is supposed to go in?

Reply:
No, this is actually intentional - though it might not be the best way to do it.

So what I am "ultimately" trying to achieve is that every file knows it's child includes all the way to the bottom, because it's CRCs etc are the CRSs of the files and its children. But when you're parsing from the top down, you don't know what the children are as you're going down through the iterations. 
However you can build up a picture of the parents, and from that you can re-parse the files to work out the children.

So what this does is, for the current file, create a parent attribute, and add the name of its parent (`x.basename`) and add the set of parents (`parsed_xml.parent.update(x.parent)`) (some duplication, but this catches the top level case where you have only one parent - the basename).